cloudimg/ms_azure.py

```diff
@@ -0,0 +1,480 @@
+
+    The filter expression is a string with format:
+    ```
+    "tag1"='value1' and "tag2"='value2' ...
+    ```
+
+    Args:
+        tags (dict): Dicionary of strings with the tags to filter
+
+    Returns:
+        String with the filter expression.
+    """
+
+    # Create the filter expression like:
+    # "tag1"='value1" and "tag2" = 'value2" ...
+    filter_expression = ""
+    for k, v in tags.items():
+        filter_expression += "\"%s\"='%s' and " % (k, v)
+    # strip the last ' and '
+    return filter_expression.rstrip(" and ")
+
+
+@attr.s
+class UploadProgress(object):
+    """
+    Responsible for storing the upload progress.
+
+    It's expected to be instantiated by a callback function during blob
+    upload phase.
+
+    Args:
+        current (float): The current uploaded size in bytes
+        total (float): The total size to upload in bytes
+        percentage (float, optional): The upload progress percentage
+    """
+
+    current = attr.ib(type=float, repr=lambda x: "%.2fMB" % (x / 1024**2))
+    total = attr.ib(type=float, repr=lambda x: "%.2fMB" % (x / 1024**2))
+    percentage = attr.ib(type=float, repr=lambda x: "%.2f%%" % (x))
+    last_seen = attr.ib(type=float, repr=False, eq=False, factory=time.time)
+
+    @percentage.default
+    def _auto_percentage(self):
+        if self.total > 0:
+            return self.current / self.total * 100.0
+        return 0
+
+
+class AzurePublishingMetadata(PublishingMetadata):
+    """
+    A collection of metadata necessary for uploading and publishing a
+    Virtual Hard Disk (VHD) to Azure.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(AzurePublishingMetadata, self).__init__(*args, **kwargs)
+
+        assert self.container, 'A container must be defined'
+        assert self.tags, 'A tag must be defined'
+
+
+class AzureService(BaseService):
+    """
+    Azure cloud provider service.
+    It's expected to be instanciated through one of its factories.
+
+    Args:
+        blob_service_client (BlobServiceClient): The blob service client to
+                                                 manipulate the storage.
+        storage_account_name (str): The storage account name for auth.
+        storage_account_key (str): The storage account key for auth.
+    """
+    # The duration of generated SAS URIs
+    SAS_EXPIRY = 365 * 3
+    # The max concurrency for blob upload
+    UPLOAD_MAX_CONCURRENCY = 16
+    # The interval in seconds between showing upload progress in logs
+    UPLOAD_LOG_INTERVAL_SECONDS = 15
+
+    def __init__(self, blob_service_client, storage_account_name,
+                 storage_account_key, **kwargs):
+
+        self.blob_service_client = blob_service_client
+        self._account_name = storage_account_name
+        self._account_key = storage_account_key
+        self._default_endpoints_protocol = kwargs.pop(
+                                                'default_endpoints_protocol',
```

Comment:
Is it intentional that these have not been documented?  With the arguments not being present in either the signature or doc string, they are effectively hidden (no way to find they exist without reading the code).  But they're still intended to be part of the public API, aren't they?

Reply:
These entries are not necessarily part of API as they only reflect the protocol to communicate with Azure storage (`https`) and the suffix URL which isn't like to change (`blob.core.windows.net`) thus I removed them and hardcoded the URL format in the respective methods.

It's very unlikely for Microsoft to change them as it would break a lot of existing applications using Azure SDK and even if they change it it's now just a matter of replacing strings inside the methods, which wouldn't break the public API.

Thanks a lot for the good catch!